### PR TITLE
First exploration: Jörmungandr NetworkLayer

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#219 

# Overview
Just exploring. WIP.

- [x] Change tip to `(SlotId, BlockId)`. Should maybe have `BlockHeader` instead.
- [ ] We need the genesis hash! (`HasGenesisHash target` or something)

# Comments

- The `undefined`s we have to pass in the tests are a bit unpleasant.
- Maybe we don't want `newtype BlockId`

### CODE REVIEW

**Comment by KtorZ** [comment]
>  Should maybe have BlockHeader instead.

Sounds like a good idea. Extend the `BlockHeader` type, and have `BlockHeader` instead.

**Comment by KtorZ** [comment]
>  We need the genesis hash! (HasGenesisHash target or something)

Why though ?

**Comment by Anviking** [comment]
> Why though ?

```haskell
initWallet = Wallet mempty mempty (SlotId 0 0, error "todo: genesisHash")
```

**Comment by KtorZ** [comment]
Good point. I guess we have our first configuration parameter then :sweat_smile:

File: lib/http-bridge/test/unit/Cardano/Wallet/HttpBridge/NetworkSpec.hs [hunk 1]
```diff
@@ -31,46 +31,46 @@ spec = do
...
                 `shouldSatisfy` all (== 104)
 
         it "should produce no blocks if start slot is after tip" $ do
-            blocks <- runExceptT $ nextBlocks network (SlotId 107 0)
+            blocks <- runExceptT $ nextBlocks network (SlotId 107 0, undefined)
             blocks `shouldBe` Right []
 
         it "should work for the first epoch" $ do
-            Right blocks <- runExceptT $ nextBlocks network (SlotId 0 0)
+            Right blocks <- runExceptT $ nextBlocks network (SlotId 0 0, undefined)
```
Comment (KtorZ) [inline lib/http-bridge/test/unit/Cardano/Wallet/HttpBridge/NetworkSpec.hs:73]: I think the idea was more like.. to update the network layer to play nicely with this. Which implies also an update in the http bridge implementation.

**Comment by Anviking** [comment]
🤔 Oops, it would be the `parentHeaderHash` of the genesis block, which at least for jormungandr is just `\NUL`s.

Originally I believed we would have the `BlockId` / `Hash "BlockHeader"` of the current block instead.

**Comment by KtorZ** [comment]
> Originally I believed we would have the BlockId / Hash "BlockHeader" of the current block instead.

Well, we don't have the "current" block header either when we initially instantiate the wallet. I wouldn't go for a `Maybe` here for it'll pollute everything else just because of the initialization. Still, it's probably a good idea to have the genesis hash somewhere available as a config param. This allows for at least checking that we are dealing with the blockchain we expect and that we agree on the initial config.

File: lib/http-bridge/test/unit/Cardano/Wallet/HttpBridge/NetworkSpec.hs (hunk 1 above)
Comment (Anviking) [reply lib/http-bridge/test/unit/Cardano/Wallet/HttpBridge/NetworkSpec.hs:73]: The idea was to have:

```haskell
data NetworkLayer t m = NetworkLayer
    { nextBlocks :: BlockHeader -> ExceptT ErrNetworkUnreachable m [Block]
    , networkTip :: ExceptT ErrNetworkTip m BlockHeader
```

And these tests use the `NetworkLayer`, not the `HttpBridgeLayer`.

Without parametrising over `tip`, I don't see how we can avoid supplying `undefined` in place of a `Hash "BlockHeader"` here.

**Comment by Anviking** [comment]
Starting from a clean slate -> #384

File: lib/core/src/Cardano/Wallet.hs [hunk 2]
```diff
@@ -461,34 +462,35 @@ newWalletLayer db nw tl = do
                 restoreSleep wid slot
             Right blocks -> do
                 let next = view #slotId . header . last $ blocks
+                let nextBlock = BlockId $ view #prevBlockHash. header . last $ blocks -- TODO: WRONG!! THIS IS PARENT!!!
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet.hs:465]: I think `nextBlocks nw` should be responsible for returning a new cursor

**Pull request closed without merging**